src/commands/force/mdapi/beta/retrieve/report.ts

```diff
@@ -0,0 +1,219 @@
+      format: 'metadata',
+      zipFileName: this.zipFileName,
+      unzip: this.unzip,
+    });
+
+    if (this.isAsync) {
+      // For async reports (wait == 0) we just check the status and call post()
+      // to write the zip file if it's done.
+      this.retrieveStatus = await this.mdapiRetrieve.checkStatus();
+      if (this.retrieveStatus.done) {
+        this.retrieveResult = await this.mdapiRetrieve.post(this.retrieveStatus);
+      }
+    } else {
+      this.ux.setSpinnerStatus(spinnerMessages.getMessage('retrieve.polling'));
+      this.retrieveResult = await this.mdapiRetrieve.pollStatus(1000, this.wait.seconds);
+    }
+
+    this.ux.stopSpinner();
+  }
+
+  // No-op implementation since any RetrieveResult status would be a success.
+  // The only time this command would report an error is if it failed
+  // flag parsing or some error during the request, and those are captured
+  // by the command framework.
+  /* eslint-disable-next-line @typescript-eslint/no-empty-function */
+  protected resolveSuccess(): void {}
+
+  protected formatResult(): RetrieveCommandResult | RetrieveCommandAsyncResult {
+    if (this.isAsync && !this.retrieveResult) {
+      this.ux.log('');
+      this.ux.log(`Retrieve Status: ${this.retrieveStatus.status}`);
+      return {
+        done: this.retrieveStatus.done,
+        id: this.retrieveStatus.id,
+        state: this.retrieveStatus.status,
+        status: this.retrieveStatus.status,
+        timedOut: true,
+      } as RetrieveCommandAsyncResult;
+    }
+
+    const formatterOptions = {
+      waitTime: this.wait.quantity,
+      verbose: this.getFlag<boolean>('verbose', false),
+      retrieveTargetDir: this.retrieveTargetDir,
+      zipFileName: this.zipFileName,
+      unzip: this.unzip,
+    };
+    const formatter = new RetrieveResultFormatter(this.logger, this.ux, formatterOptions, this.retrieveResult);
+
+    if (!this.isJsonOutput()) {
+      formatter.display();
+    }
+    return formatter.getJson();
+  }
+
+  private resolveRetrieveId(stash?: MdRetrieveData): string {
+    const retrieveId = this.getFlag<string>('jobid', stash?.jobid);
+    // throw if no Retrieve ID
+    if (!retrieveId) {
+      throw SfdxError.create('@salesforce/plugin-source', 'md.retrieve', 'MissingRetrieveId');
+    }
+    return retrieveId;
+  }
+
+  private ensureFlagPath(options: EnsureFlagOptions): string {
+    const { flagName, path, type, throwOnENOENT } = options;
+
+    const trimmedPath = path?.trim();
+    let resolvedPath: string;
+    if (trimmedPath?.length) {
+      resolvedPath = resolve(trimmedPath);
+    }
+
+    try {
+      const stats = fs.statSync(resolvedPath);
+      if (type !== 'any') {
+        const isDir = stats.isDirectory();
+        if (type === 'dir' && !isDir) {
+          const msg = 'Expected a directory but found a file';
+          throw SfdxError.create('@salesforce/plugin-source', 'md.convert', 'InvalidFlagPath', [flagName, path, msg]);
```

Comment:
several of these are referencing md.convert messages.  If that was intentional, you'll need to load them, too.

Reply:
I think when you wrote these, they made sense.  As they're being used more, I think it'd be easier to 
1. make flag validation functions to use with the flags' `parse` options like `dirMustExist()` or `ensureDirExists`
2. OR make some custom flags like `directoryThatExists` or `ensuredDirectory`

I understand that probably goes in the consolidation PR but these are some opinions.  I'd really love to get these into the sfdxFlags (ex: `flags.directory` would have some options to do these types of things)

Reply:
This was all for the consolidation PR to follow.

I was thinking about using flag validation as I saw how this was starting to unfold.  I think that's a better, more oclif way of doing this so that's probably the approach I'll take but I'm not 100% sure since sometimes it's ok that the directory doesn't exist because (e.g.) the command will then use the default package dir.  I'll find out soon.  :)

I'm not planning on keeping the messages in the convert message bundle, but you don't have to load them explicitly when you use the message in `SfdxError.create`.  It handles it for you.

Reply:
OH!